tests/resource/aws_synthetics_canary: Canaries entering ERROR state

<!--- Please keep this note for the community --->

### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

<!--- Thank you for keeping this note for the community --->

### Description

The `aws_synthetics_canary` testing is consistently failing in ways that indicate that the Synthetic Canaries are creating "successfully" but then entering an `ERROR` state. This includes the `engine_arn` being removed, `Canary is not in a startable state` / `Canary in a state that can't be modified` type errors, and potentially `vpc_config` being removed.

```
=== CONT  TestAccAWSSyntheticsCanary_basic
resource_aws_synthetics_canary_test.go:82: Step 1/3 error: Check failed: 2 errors occurred:
  * Check 14/20 error: aws_synthetics_canary.test: Attribute 'engine_arn' didn't match "arn:aws:lambda:us-west-2:*******:function:cwsyn-tf-acc-test-96pbyi8n.+", got ""
  * Check 20/20 error: aws_synthetics_canary.test: Attribute 'status' expected "READY", got "ERROR"
--- FAIL: TestAccAWSSyntheticsCanary_basic (17.20s)

=== CONT  TestAccAWSSyntheticsCanary_startCanary
resource_aws_synthetics_canary_test.go:190: Step 1/4 error: Error running apply: exit status 1
Error: error starting Synthetics Canary: ConflictException: Canary is not in a startable state.
on terraform_plugin_test.tf line 92, in resource "aws_synthetics_canary" "test":
92: resource "aws_synthetics_canary" "test" {
--- FAIL: TestAccAWSSyntheticsCanary_startCanary (19.68s)

=== CONT  TestAccAWSSyntheticsCanary_startCanary_codeChanges
resource_aws_synthetics_canary_test.go:238: Step 1/3 error: Error running apply: exit status 1
Error: error starting Synthetics Canary: ConflictException: Canary is not in a startable state.
on terraform_plugin_test.tf line 92, in resource "aws_synthetics_canary" "test":
92: resource "aws_synthetics_canary" "test" {
--- FAIL: TestAccAWSSyntheticsCanary_startCanary_codeChanges (21.37s)

=== CONT  TestAccAWSSyntheticsCanary_runtimeVersion
resource_aws_synthetics_canary_test.go:155: Step 3/3 error: Error running apply: exit status 1
Error: error updating Synthetics Canary: ConflictException: Canary in a state that can't be modified: ERROR
on terraform_plugin_test.tf line 92, in resource "aws_synthetics_canary" "test":
92: resource "aws_synthetics_canary" "test" {
--- FAIL: TestAccAWSSyntheticsCanary_runtimeVersion (40.53s)

=== CONT  TestAccAWSSyntheticsCanary_s3
resource_aws_synthetics_canary_test.go:279: Step 1/2 error: Check failed: 1 error occurred:
  * Check 15/18 error: aws_synthetics_canary.test: Attribute 'engine_arn' didn't match "arn:aws:lambda:us-west-2:*******:function:cwsyn-tf-acc-test-aacojbil.+", got ""
--- FAIL: TestAccAWSSyntheticsCanary_s3 (23.62s)

=== CONT  TestAccAWSSyntheticsCanary_vpc
resource_aws_synthetics_canary_test.go:411: Step 1/4 error: Check failed: 3 errors occurred:
  * Check 2/4 error: aws_synthetics_canary.test: Attribute 'vpc_config.0.subnet_ids.#' not found
  * Check 3/4 error: aws_synthetics_canary.test: Attribute 'vpc_config.0.security_group_ids.#' not found
  * Check 4/4 error: aws_synthetics_canary.test: Attribute "vpc_config.0.vpc_id" not set, but "id" is set in aws_vpc.test as "vpc-0e1516c8f69f40513"
--- FAIL: TestAccAWSSyntheticsCanary_vpc (20.51s)

=== CONT  TestAccAWSSyntheticsCanary_runConfig
resource_aws_synthetics_canary_test.go:323: Step 3/4 error: Error running apply: exit status 1
Error: error updating Synthetics Canary: ConflictException: Canary in a state that can't be modified: ERROR
on terraform_plugin_test.tf line 92, in resource "aws_synthetics_canary" "test":
92: resource "aws_synthetics_canary" "test" {
--- FAIL: TestAccAWSSyntheticsCanary_runConfig (37.71s)

=== CONT  TestAccAWSSyntheticsCanary_runConfigTracing
resource_aws_synthetics_canary_test.go:369: Step 3/4 error: Error running apply: exit status 1
Error: error updating Synthetics Canary: ConflictException: Canary in a state that can't be modified: ERROR
on terraform_plugin_test.tf line 92, in resource "aws_synthetics_canary" "test":
92: resource "aws_synthetics_canary" "test" {
--- FAIL: TestAccAWSSyntheticsCanary_runConfigTracing (37.42s)
```

### Affected Resource

* aws_synthetics_canary

--- a/.changelog/20463.txt
+++ b/.changelog/20463.txt
@@ -1,0 +1,3 @@
+```release-note:bug
+resource/aws_synthetics_canary: Correctly report any resource creation errors
+```

--- a/aws/internal/service/synthetics/finder/finder.go
+++ b/aws/internal/service/synthetics/finder/finder.go
@@ -3,18 +3,34 @@ package finder
 import (
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/synthetics"
+	"github.com/hashicorp/aws-sdk-go-base/tfawserr"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 )
 
-// CanaryByName returns the Canary corresponding to the specified Name.
-func CanaryByName(conn *synthetics.Synthetics, name string) (*synthetics.GetCanaryOutput, error) {
+func CanaryByName(conn *synthetics.Synthetics, name string) (*synthetics.Canary, error) {
 	input := &synthetics.GetCanaryInput{
 		Name: aws.String(name),
 	}
 
 	output, err := conn.GetCanary(input)
+
+	if tfawserr.ErrCodeEquals(err, synthetics.ErrCodeResourceNotFoundException) {
+		return nil, &resource.NotFoundError{
+			LastError:   err,
+			LastRequest: input,
+		}
+	}
+
 	if err != nil {
 		return nil, err
 	}
 
-	return output, nil
+	if output == nil || output.Canary == nil || output.Canary.Status == nil {
+		return nil, &resource.NotFoundError{
+			Message:     "Empty result",
+			LastRequest: input,
+		}
+	}
+
+	return output.Canary, nil
 }

--- a/aws/internal/service/synthetics/waiter/status.go
+++ b/aws/internal/service/synthetics/waiter/status.go
@@ -1,26 +1,25 @@
 package waiter
 
 import (
-	"fmt"
-
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/synthetics"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 	"github.com/terraform-providers/terraform-provider-aws/aws/internal/service/synthetics/finder"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/tfresource"
 )
 
-func CanaryStatus(conn *synthetics.Synthetics, name string) resource.StateRefreshFunc {
+func CanaryState(conn *synthetics.Synthetics, name string) resource.StateRefreshFunc {
 	return func() (interface{}, string, error) {
 		output, err := finder.CanaryByName(conn, name)
 
+		if tfresource.NotFound(err) {
+			return nil, "", nil
+		}
+
 		if err != nil {
-			return nil, synthetics.CanaryStateError, err
+			return nil, "", err
 		}
 
-		if aws.StringValue(output.Canary.Status.State) == synthetics.CanaryStateError {
-			return output, synthetics.CanaryStateError, fmt.Errorf("%s: %s", aws.StringValue(output.Canary.Status.StateReasonCode), aws.StringValue(output.Canary.Status.StateReason))
-		}
-
-		return output, aws.StringValue(output.Canary.Status.State), nil
+		return output, aws.StringValue(output.Status.State), nil
 	}
 }

--- a/aws/internal/service/synthetics/waiter/waiter.go
+++ b/aws/internal/service/synthetics/waiter/waiter.go
@@ -1,89 +1,112 @@
 package waiter
 
 import (
+	"fmt"
 	"time"
 
+	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/service/synthetics"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/tfresource"
 )
 
 const (
-	// Maximum amount of time to wait for a Canary to return Ready
 	CanaryCreatedTimeout = 5 * time.Minute
 	CanaryRunningTimeout = 5 * time.Minute
 	CanaryStoppedTimeout = 5 * time.Minute
 	CanaryDeletedTimeout = 5 * time.Minute
 )
 
-// CanaryReady waits for a Canary to return Ready
-func CanaryReady(conn *synthetics.Synthetics, name string) (*synthetics.GetCanaryOutput, error) {
+func CanaryReady(conn *synthetics.Synthetics, name string) (*synthetics.Canary, error) {
 	stateConf := &resource.StateChangeConf{
 		Pending: []string{synthetics.CanaryStateCreating, synthetics.CanaryStateUpdating},
 		Target:  []string{synthetics.CanaryStateReady},
-		Refresh: CanaryStatus(conn, name),
+		Refresh: CanaryState(conn, name),
 		Timeout: CanaryCreatedTimeout,
 	}
 
 	outputRaw, err := stateConf.WaitForState()
 
-	if v, ok := outputRaw.(*synthetics.GetCanaryOutput); ok {
-		return v, err
+	if output, ok := outputRaw.(*synthetics.Canary); ok {
+		if status := output.Status; aws.StringValue(status.State) == synthetics.CanaryStateError {
+			tfresource.SetLastError(err, fmt.Errorf("%s: %s", aws.StringValue(status.StateReasonCode), aws.StringValue(status.StateReason)))
+		}
+
+		return output, err
 	}
 
 	return nil, err
 }
 
-// CanaryReady waits for a Canary to return Stopped
-func CanaryStopped(conn *synthetics.Synthetics, name string) (*synthetics.GetCanaryOutput, error) {
+func CanaryStopped(conn *synthetics.Synthetics, name string) (*synthetics.Canary, error) {
 	stateConf := &resource.StateChangeConf{
-		Pending: []string{synthetics.CanaryStateStopping, synthetics.CanaryStateUpdating,
-			synthetics.CanaryStateRunning, synthetics.CanaryStateReady, synthetics.CanaryStateStarting},
+		Pending: []string{
+			synthetics.CanaryStateStopping,
+			synthetics.CanaryStateUpdating,
+			synthetics.CanaryStateRunning,
+			synthetics.CanaryStateReady,
+			synthetics.CanaryStateStarting,
+		},
 		Target:  []string{synthetics.CanaryStateStopped},
-		Refresh: CanaryStatus(conn, name),
+		Refresh: CanaryState(conn, name),
 		Timeout: CanaryStoppedTimeout,
 	}
 
 	outputRaw, err := stateConf.WaitForState()
 
-	if v, ok := outputRaw.(*synthetics.GetCanaryOutput); ok {
-		return v, err
+	if output, ok := outputRaw.(*synthetics.Canary); ok {
+		if status := output.Status; aws.StringValue(status.State) == synthetics.CanaryStateError {
+			tfresource.SetLastError(err, fmt.Errorf("%s: %s", aws.StringValue(status.StateReasonCode), aws.StringValue(status.StateReason)))
+		}
+
+		return output, err
 	}
 
 	return nil, err
 }
 
-// CanaryReady waits for a Canary to return Running
-func CanaryRunning(conn *synthetics.Synthetics, name string) (*synthetics.GetCanaryOutput, error) {
+func CanaryRunning(conn *synthetics.Synthetics, name string) (*synthetics.Canary, error) {
 	stateConf := &resource.StateChangeConf{
-		Pending: []string{synthetics.CanaryStateStarting, synthetics.CanaryStateUpdating,
-			synthetics.CanaryStateStarting, synthetics.CanaryStateReady},
+		Pending: []string{
+			synthetics.CanaryStateStarting,
+			synthetics.CanaryStateUpdating,
+			synthetics.CanaryStateStarting,
+			synthetics.CanaryStateReady,
+		},
 		Target:  []string{synthetics.CanaryStateRunning},
-		Refresh: CanaryStatus(conn, name),
+		Refresh: CanaryState(conn, name),
 		Timeout: CanaryRunningTimeout,
 	}
 
 	outputRaw, err := stateConf.WaitForState()
 
-	if v, ok := outputRaw.(*synthetics.GetCanaryOutput); ok {
-		return v, err
+	if output, ok := outputRaw.(*synthetics.Canary); ok {
+		if status := output.Status; aws.StringValue(status.State) == synthetics.CanaryStateError {
+			tfresource.SetLastError(err, fmt.Errorf("%s: %s", aws.StringValue(status.StateReasonCode), aws.StringValue(status.StateReason)))
+		}
+
+		return output, err
 	}
 
 	return nil, err
 }
 
-// CanaryReady waits for a Canary to return Ready
-func CanaryDeleted(conn *synthetics.Synthetics, name string) (*synthetics.GetCanaryOutput, error) {
+func CanaryDeleted(conn *synthetics.Synthetics, name string) (*synthetics.Canary, error) {
 	stateConf := &resource.StateChangeConf{
 		Pending: []string{synthetics.CanaryStateDeleting, synthetics.CanaryStateStopped},
 		Target:  []string{},
-		Refresh: CanaryStatus(conn, name),
+		Refresh: CanaryState(conn, name),
 		Timeout: CanaryDeletedTimeout,
 	}
 
 	outputRaw, err := stateConf.WaitForState()
 
-	if v, ok := outputRaw.(*synthetics.GetCanaryOutput); ok {
-		return v, err
+	if output, ok := outputRaw.(*synthetics.Canary); ok {
+		if status := output.Status; aws.StringValue(status.State) == synthetics.CanaryStateError {
+			tfresource.SetLastError(err, fmt.Errorf("%s: %s", aws.StringValue(status.StateReasonCode), aws.StringValue(status.StateReason)))
+		}
+
+		return output, err
 	}
 
 	return nil, err

--- a/aws/resource_aws_synthetics_canary.go
+++ b/aws/resource_aws_synthetics_canary.go
@@ -10,7 +10,7 @@ import (
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/arn"
 	"github.com/aws/aws-sdk-go/service/synthetics"
-	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/hashicorp/aws-sdk-go-base/tfawserr"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
 	"github.com/terraform-providers/terraform-provider-aws/aws/internal/keyvaluetags"
@@ -227,23 +227,21 @@ func resourceAwsSyntheticsCanaryCreate(d *schema.ResourceData, meta interface{})
 	defaultTagsConfig := meta.(*AWSClient).DefaultTagsConfig
 	tags := defaultTagsConfig.MergeTags(keyvaluetags.New(d.Get("tags").(map[string]interface{})))
 
+	name := d.Get("name").(string)
 	input := &synthetics.CreateCanaryInput{
-		Name:               aws.String(d.Get("name").(string)),
 		ArtifactS3Location: aws.String(d.Get("artifact_s3_location").(string)),
 		ExecutionRoleArn:   aws.String(d.Get("execution_role_arn").(string)),
+		Name:               aws.String(name),
 		RuntimeVersion:     aws.String(d.Get("runtime_version").(string)),
 	}
 
 	code, err := expandAwsSyntheticsCanaryCode(d)
+
 	if err != nil {
 		return err
 	}
 
 	input.Code = code
-
-	if len(tags) > 0 {
-		input.Tags = tags.IgnoreAws().SyntheticsTags()
-	}
 
 	if v, ok := d.GetOk("run_config"); ok {
 		input.RunConfig = expandAwsSyntheticsCanaryRunConfig(v.([]interface{}))
@@ -265,7 +263,18 @@ func resourceAwsSyntheticsCanaryCreate(d *schema.ResourceData, meta interface{})
 		input.SuccessRetentionPeriodInDays = aws.Int64(int64(v.(int)))
 	}
 
-	log.Printf("[DEBUG] creating Synthetics Canary: %#v", input)
+	if len(tags) > 0 {
+		input.Tags = tags.IgnoreAws().SyntheticsTags()
+	}
+
+	log.Printf("[DEBUG] Creating Synthetics Canary: %s", input)
+	output, err := conn.CreateCanary(input)
+
+	if err != nil {
+		return fmt.Errorf("error creating Synthetics Canary (%s): %w", name, err)
+	}
+
+	d.SetId(aws.StringValue(output.Canary.Name))
 
 	// Underlying IAM eventual consistency errors can occur after the creation
 	// operation. The goal is only retry these types of errors up to the IAM
@@ -273,56 +282,27 @@ func resourceAwsSyntheticsCanaryCreate(d *schema.ResourceData, meta interface{})
 	// its own timeout, we store a stop time upfront for checking.
 	iamwaiterStopTime := time.Now().Add(iamwaiter.PropagationTimeout)
 
-	// Ensure to add IAM eventual consistency timeout in case of retries
-	err = resource.Retry(iamwaiter.PropagationTimeout+waiter.CanaryCreatedTimeout, func() *resource.RetryError {
-		// Only retry IAM eventual consistency errors up to that timeout
-		iamwaiterRetry := time.Now().Before(iamwaiterStopTime)
-
-		resp, err := conn.CreateCanary(input)
-
-		if err != nil {
-			return resource.NonRetryableError(fmt.Errorf("error creating Synthetics Canary: %w", err))
-		}
-
-		if resp == nil || resp.Canary == nil {
-			return resource.NonRetryableError(fmt.Errorf("error creating Synthetics Canary: empty response"))
-		}
-
-		d.SetId(aws.StringValue(resp.Canary.Name))
-
-		_, err = waiter.CanaryReady(conn, d.Id())
-
-		if err != nil {
-			// This error synthesized from the Status object and not an AWS SDK Go error type
-			if iamwaiterRetry && strings.Contains(err.Error(), "The role defined for the function cannot be assumed by Lambda") {
-				return resource.RetryableError(fmt.Errorf("error waiting for Synthetics Canary (%s) creation: %w", d.Id(), err))
+	_, err = tfresource.RetryWhen(
+		iamwaiter.PropagationTimeout+waiter.CanaryCreatedTimeout,
+		func() (interface{}, error) {
+			return waiter.CanaryReady(conn, d.Id())
+		},
+		func(err error) bool {
+			// Only retry IAM eventual consistency errors up to that timeout.
+			if err != nil && time.Now().Before(iamwaiterStopTime) {
+				// This error synthesized from the Status object and not an AWS SDK Go error type.
+				return strings.Contains(err.Error(), "The role defined for the function cannot be assumed by Lambda")
 			}
 
-			return resource.NonRetryableError(fmt.Errorf("error waiting for Synthetics Canary (%s) creation: %w", d.Id(), err))
-		}
+			return false
+		},
+	)
 
-		return nil
-	})
-
-	if tfresource.TimedOut(err) {
-		resp, err := conn.CreateCanary(input)
-
-		if err != nil {
-			return fmt.Errorf("error creating Synthetics Canary: %w", err)
-		}
-
-		if resp == nil || resp.Canary == nil {
-			return fmt.Errorf("error creating Synthetics Canary: empty response")
-		}
-
-		d.SetId(aws.StringValue(resp.Canary.Name))
-
-		if _, err = waiter.CanaryReady(conn, d.Id()); err != nil {
-			return fmt.Errorf("error waiting for Synthetics Canary (%s) creation: %w", d.Id(), err)
-		}
+	if err != nil {
+		return fmt.Errorf("error waiting for Synthetics Canary (%s) create: %w", d.Id(), err)
 	}
 
-	if v := d.Get("start_canary"); v.(bool) {
+	if d.Get("start_canary").(bool) {
 		if err := syntheticsStartCanary(d.Id(), conn); err != nil {
 			return err
 		}
@@ -336,17 +316,26 @@ func resourceAwsSyntheticsCanaryRead(d *schema.ResourceData, meta interface{}) e
 	defaultTagsConfig := meta.(*AWSClient).DefaultTagsConfig
 	ignoreTagsConfig := meta.(*AWSClient).IgnoreTagsConfig
 
-	resp, err := finder.CanaryByName(conn, d.Id())
-	if err != nil {
-		if isAWSErr(err, synthetics.ErrCodeResourceNotFoundException, "") {
-			log.Printf("[WARN] Synthetics Canary (%s) not found, removing from state", d.Id())
-			d.SetId("")
-			return nil
-		}
-		return fmt.Errorf("error reading Synthetics Canary: %w", err)
+	canary, err := finder.CanaryByName(conn, d.Id())
+
+	if !d.IsNewResource() && tfresource.NotFound(err) {
+		log.Printf("[WARN] Synthetics Canary (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
 	}
 
-	canary := resp.Canary
+	if err != nil {
+		return fmt.Errorf("error reading Synthetics Canary (%s): %w", d.Id(), err)
+	}
+
+	canaryArn := arn.ARN{
+		Partition: meta.(*AWSClient).partition,
+		Service:   synthetics.ServiceName,
+		Region:    meta.(*AWSClient).region,
+		AccountID: meta.(*AWSClient).accountid,
+		Resource:  fmt.Sprintf("canary:%s", aws.StringValue(canary.Name)),
+	}.String()
+	d.Set("arn", canaryArn)
 	d.Set("artifact_s3_location", canary.ArtifactS3Location)
 	d.Set("engine_arn", canary.EngineArn)
 	d.Set("execution_role_arn", canary.ExecutionRoleArn)
@@ -357,16 +346,6 @@ func resourceAwsSyntheticsCanaryRead(d *schema.ResourceData, meta interface{}) e
 	d.Set("source_location_arn", canary.Code.SourceLocationArn)
 	d.Set("status", canary.Status.State)
 	d.Set("success_retention_period", canary.SuccessRetentionPeriodInDays)
-
-	canaryArn := arn.ARN{
-		Partition: meta.(*AWSClient).partition,
-		Service:   synthetics.ServiceName,
-		Region:    meta.(*AWSClient).region,
-		AccountID: meta.(*AWSClient).accountid,
-		Resource:  fmt.Sprintf("canary:%s", aws.StringValue(canary.Name)),
-	}.String()
-
-	d.Set("arn", canaryArn)
 
 	if err := d.Set("vpc_config", flattenAwsSyntheticsCanaryVpcConfig(canary.VpcConfig)); err != nil {
 		return fmt.Errorf("error setting vpc config: %w", err)
@@ -452,22 +431,24 @@ func resourceAwsSyntheticsCanaryUpdate(d *schema.ResourceData, meta interface{})
 			}
 		}
 
+		log.Printf("[DEBUG] Updating Synthetics Canary: %s", input)
 		_, err := conn.UpdateCanary(input)
+
 		if err != nil {
-			return fmt.Errorf("error updating Synthetics Canary: %w", err)
+			return fmt.Errorf("error updating Synthetics Canary (%s): %w", d.Id(), err)
 		}
 
 		if status != synthetics.CanaryStateReady {
 			if _, err := waiter.CanaryStopped(conn, d.Id()); err != nil {
-				return fmt.Errorf("error waiting for Synthetics Canary (%s) to be stopped: %w", d.Id(), err)
+				return fmt.Errorf("error waiting for Synthetics Canary (%s) stop: %w", d.Id(), err)
 			}
 		} else {
 			if _, err := waiter.CanaryReady(conn, d.Id()); err != nil {
-				return fmt.Errorf("error waiting for Synthetics Canary (%s) updating: %w", d.Id(), err)
+				return fmt.Errorf("error waiting for Synthetics Canary (%s) ready: %w", d.Id(), err)
 			}
 		}
 
-		if v := d.Get("start_canary"); v.(bool) {
+		if d.Get("start_canary").(bool) {
 			if err := syntheticsStartCanary(d.Id(), conn); err != nil {
 				return err
 			}
@@ -476,7 +457,7 @@ func resourceAwsSyntheticsCanaryUpdate(d *schema.ResourceData, meta interface{})
 
 	if d.HasChange("start_canary") {
 		status := d.Get("status").(string)
-		if v := d.Get("start_canary"); v.(bool) {
+		if d.Get("start_canary").(bool) {
 			if status != synthetics.CanaryStateRunning {
 				if err := syntheticsStartCanary(d.Id(), conn); err != nil {
 					return err
@@ -505,29 +486,29 @@ func resourceAwsSyntheticsCanaryUpdate(d *schema.ResourceData, meta interface{})
 func resourceAwsSyntheticsCanaryDelete(d *schema.ResourceData, meta interface{}) error {
 	conn := meta.(*AWSClient).syntheticsconn
 
-	if status := d.Get("status"); status.(string) == synthetics.CanaryStateRunning {
+	if status := d.Get("status").(string); status == synthetics.CanaryStateRunning {
 		if err := syntheticsStopCanary(d.Id(), conn); err != nil {
-			log.Print("could not stop canary before delete")
+			return err
 		}
 	}
 
-	input := &synthetics.DeleteCanaryInput{
+	log.Printf("[DEBUG] Deleting Synthetics Canary: (%s)", d.Id())
+	_, err := conn.DeleteCanary(&synthetics.DeleteCanaryInput{
 		Name: aws.String(d.Id()),
+	})
+
+	if tfawserr.ErrCodeEquals(err, synthetics.ErrCodeResourceNotFoundException) {
+		return nil
 	}
 
-	_, err := conn.DeleteCanary(input)
 	if err != nil {
-		if isAWSErr(err, synthetics.ErrCodeResourceNotFoundException, "") {
-			return nil
-		}
-		return fmt.Errorf("error deleting Synthetics Canary: %w", err)
+		return fmt.Errorf("error deleting Synthetics Canary (%s): %w", d.Id(), err)
 	}
 
-	if _, err := waiter.CanaryDeleted(conn, d.Id()); err != nil {
-		if isAWSErr(err, synthetics.ErrCodeResourceNotFoundException, "") {
-			return nil
-		}
-		return fmt.Errorf("error waiting for Synthetics Canary (%s) deletion: %w", d.Id(), err)
+	_, err = waiter.CanaryDeleted(conn, d.Id())
+
+	if err != nil {
+		return fmt.Errorf("error waiting for Synthetics Canary (%s) delete: %w", d.Id(), err)
 	}
 
 	return nil
@@ -679,35 +660,38 @@ func flattenAwsSyntheticsCanaryTimeline(timeline *synthetics.CanaryTimeline) []i
 }
 
 func syntheticsStartCanary(name string, conn *synthetics.Synthetics) error {
-	startInput := &synthetics.StartCanaryInput{
+	log.Printf("[DEBUG] Starting Synthetics Canary: (%s)", name)
+	_, err := conn.StartCanary(&synthetics.StartCanaryInput{
 		Name: aws.String(name),
-	}
+	})
 
-	log.Print("starting Canary")
-	_, err := conn.StartCanary(startInput)
 	if err != nil {
-		return fmt.Errorf("error starting Synthetics Canary: %w", err)
+		return fmt.Errorf("error starting Synthetics Canary (%s): %w", name, err)
 	}
 
-	if _, err := waiter.CanaryRunning(conn, name); err != nil {
-		return fmt.Errorf("error waiting for Synthetics Canary (%s) to be running: %w", name, err)
+	_, err = waiter.CanaryRunning(conn, name)
+
+	if err != nil {
+		return fmt.Errorf("error waiting for Synthetics Canary (%s) start: %w", name, err)
 	}
 
 	return nil
 }
 
 func syntheticsStopCanary(name string, conn *synthetics.Synthetics) error {
-	stopInput := &synthetics.StopCanaryInput{
+	log.Printf("[DEBUG] Stopping Synthetics Canary: (%s)", name)
+	_, err := conn.StopCanary(&synthetics.StopCanaryInput{
 		Name: aws.String(name),
-	}
+	})
 
-	_, err := conn.StopCanary(stopInput)
 	if err != nil {
-		return fmt.Errorf("error stopping Synthetics Canary: %w", err)
+		return fmt.Errorf("error stopping Synthetics Canary (%s): %w", name, err)
 	}
 
-	if _, err := waiter.CanaryStopped(conn, name); err != nil {
-		return fmt.Errorf("error waiting for Synthetics Canary (%s) to be stopped: %w", name, err)
+	_, err = waiter.CanaryStopped(conn, name)
+
+	if err != nil {
+		return fmt.Errorf("error waiting for Synthetics Canary (%s) stop: %w", name, err)
 	}
 
 	return nil

--- a/aws/resource_aws_synthetics_canary_test.go
+++ b/aws/resource_aws_synthetics_canary_test.go
@@ -92,7 +92,7 @@ func TestAccAWSSyntheticsCanary_basic(t *testing.T) {
 					testAccCheckAwsSyntheticsCanaryExists(resourceName, &conf1),
 					testAccMatchResourceAttrRegionalARN(resourceName, "arn", synthetics.ServiceName, regexp.MustCompile(`canary:.+`)),
 					resource.TestCheckResourceAttr(resourceName, "name", rName),
-					resource.TestCheckResourceAttr(resourceName, "runtime_version", "syn-1.0"),
+					resource.TestCheckResourceAttr(resourceName, "runtime_version", "syn-nodejs-puppeteer-3.2"),
 					resource.TestCheckResourceAttr(resourceName, "tags.%", "0"),
 					resource.TestCheckResourceAttr(resourceName, "run_config.0.memory_in_mb", "1000"),
 					resource.TestCheckResourceAttr(resourceName, "run_config.0.timeout_in_seconds", "840"),
@@ -123,7 +123,7 @@ func TestAccAWSSyntheticsCanary_basic(t *testing.T) {
 					testAccCheckAwsSyntheticsCanaryExists(resourceName, &conf2),
 					testAccMatchResourceAttrRegionalARN(resourceName, "arn", synthetics.ServiceName, regexp.MustCompile(`canary:.+`)),
 					resource.TestCheckResourceAttr(resourceName, "name", rName),
-					resource.TestCheckResourceAttr(resourceName, "runtime_version", "syn-1.0"),
+					resource.TestCheckResourceAttr(resourceName, "runtime_version", "syn-nodejs-puppeteer-3.2"),
 					resource.TestCheckResourceAttr(resourceName, "tags.%", "0"),
 					resource.TestCheckResourceAttr(resourceName, "run_config.0.memory_in_mb", "1000"),
 					resource.TestCheckResourceAttr(resourceName, "run_config.0.timeout_in_seconds", "840"),
@@ -160,10 +160,10 @@ func TestAccAWSSyntheticsCanary_runtimeVersion(t *testing.T) {
 		CheckDestroy: testAccCheckAwsSyntheticsCanaryDestroy,
 		Steps: []resource.TestStep{
 			{
-				Config: testAccAWSSyntheticsCanaryRuntimeVersionConfig(rName, "syn-nodejs-2.1"),
+				Config: testAccAWSSyntheticsCanaryRuntimeVersionConfig(rName, "syn-nodejs-puppeteer-3.1"),
 				Check: resource.ComposeAggregateTestCheckFunc(
 					testAccCheckAwsSyntheticsCanaryExists(resourceName, &conf1),
-					resource.TestCheckResourceAttr(resourceName, "runtime_version", "syn-nodejs-2.1"),
+					resource.TestCheckResourceAttr(resourceName, "runtime_version", "syn-nodejs-puppeteer-3.1"),
 				),
 			},
 			{
@@ -173,10 +173,10 @@ func TestAccAWSSyntheticsCanary_runtimeVersion(t *testing.T) {
 				ImportStateVerifyIgnore: []string{"zip_file", "start_canary"},
 			},
 			{
-				Config: testAccAWSSyntheticsCanaryRuntimeVersionConfig(rName, "syn-nodejs-2.2"),
+				Config: testAccAWSSyntheticsCanaryRuntimeVersionConfig(rName, "syn-nodejs-puppeteer-3.2"),
 				Check: resource.ComposeAggregateTestCheckFunc(
 					testAccCheckAwsSyntheticsCanaryExists(resourceName, &conf1),
-					resource.TestCheckResourceAttr(resourceName, "runtime_version", "syn-nodejs-2.2"),
+					resource.TestCheckResourceAttr(resourceName, "runtime_version", "syn-nodejs-puppeteer-3.2"),
 				),
 			},
 		},
@@ -289,7 +289,7 @@ func TestAccAWSSyntheticsCanary_s3(t *testing.T) {
 					testAccCheckAwsSyntheticsCanaryExists(resourceName, &conf),
 					testAccMatchResourceAttrRegionalARN(resourceName, "arn", synthetics.ServiceName, regexp.MustCompile(`canary:.+`)),
 					resource.TestCheckResourceAttr(resourceName, "name", rName),
-					resource.TestCheckResourceAttr(resourceName, "runtime_version", "syn-1.0"),
+					resource.TestCheckResourceAttr(resourceName, "runtime_version", "syn-nodejs-puppeteer-3.2"),
 					resource.TestCheckResourceAttr(resourceName, "tags.%", "0"),
 					resource.TestCheckResourceAttr(resourceName, "run_config.0.memory_in_mb", "1000"),
 					resource.TestCheckResourceAttr(resourceName, "run_config.0.timeout_in_seconds", "840"),
@@ -746,7 +746,7 @@ resource "aws_synthetics_canary" "test" {
   execution_role_arn   = aws_iam_role.test.arn
   handler              = "exports.handler"
   zip_file             = "test-fixtures/lambdatest.zip"
-  runtime_version      = "syn-1.0"
+  runtime_version      = "syn-nodejs-puppeteer-3.2"
 
   schedule {
     expression = "rate(0 minute)"
@@ -767,7 +767,7 @@ resource "aws_synthetics_canary" "test" {
   execution_role_arn   = aws_iam_role.test.arn
   handler              = "exports.handler"
   zip_file             = "test-fixtures/lambdatest.zip"
-  runtime_version      = "syn-1.0"
+  runtime_version      = "syn-nodejs-puppeteer-3.2"
 
   schedule {
     expression = "rate(0 minute)"
@@ -789,7 +789,7 @@ resource "aws_synthetics_canary" "test" {
   execution_role_arn   = aws_iam_role.test.arn
   handler              = "exports.handler"
   zip_file             = "test-fixtures/lambdatest.zip"
-  runtime_version      = "syn-nodejs-2.0"
+  runtime_version      = "syn-nodejs-puppeteer-3.2"
 
   schedule {
     expression = "rate(0 minute)"
@@ -811,7 +811,7 @@ resource "aws_synthetics_canary" "test" {
   execution_role_arn   = aws_iam_role.test.arn
   handler              = "exports.handler"
   zip_file             = "test-fixtures/lambdatest.zip"
-  runtime_version      = "syn-1.0"
+  runtime_version      = "syn-nodejs-puppeteer-3.2"
 
   schedule {
     expression = "rate(0 minute)"
@@ -845,7 +845,7 @@ resource "aws_synthetics_canary" "test" {
   execution_role_arn   = aws_iam_role.test.arn
   handler              = "exports.handler"
   zip_file             = "test-fixtures/lambdatest_modified.zip"
-  runtime_version      = "syn-1.0"
+  runtime_version      = "syn-nodejs-puppeteer-3.2"
 
   schedule {
     expression = "rate(0 minute)"
@@ -863,7 +863,7 @@ resource "aws_synthetics_canary" "test" {
   handler              = "exports.handler"
   zip_file             = "test-fixtures/lambdatest.zip"
   start_canary         = %[2]t
-  runtime_version      = "syn-1.0"
+  runtime_version      = "syn-nodejs-puppeteer-3.2"
 
   schedule {
     expression = "rate(0 minute)"
@@ -881,7 +881,7 @@ resource "aws_synthetics_canary" "test" {
   handler              = "exports.handler"
   zip_file             = "test-fixtures/lambdatest_modified.zip"
   start_canary         = %[2]t
-  runtime_version      = "syn-1.0"
+  runtime_version      = "syn-nodejs-puppeteer-3.2"
 
   schedule {
     expression = "rate(0 minute)"
@@ -900,7 +900,7 @@ resource "aws_synthetics_canary" "test" {
   s3_bucket            = aws_s3_bucket_object.test.bucket
   s3_key               = aws_s3_bucket_object.test.key
   s3_version           = aws_s3_bucket_object.test.version_id
-  runtime_version      = "syn-1.0"
+  runtime_version      = "syn-nodejs-puppeteer-3.2"
 
   schedule {
     expression = "rate(0 minute)"
@@ -981,7 +981,7 @@ resource "aws_synthetics_canary" "test" {
   execution_role_arn   = aws_iam_role.test.arn
   handler              = "exports.handler"
   zip_file             = "test-fixtures/lambdatest.zip"
-  runtime_version      = "syn-1.0"
+  runtime_version      = "syn-nodejs-puppeteer-3.2"
 
   schedule {
     expression = "rate(0 minute)"
@@ -1008,7 +1008,7 @@ resource "aws_synthetics_canary" "test" {
   execution_role_arn   = aws_iam_role.test.arn
   handler              = "exports.handler"
   zip_file             = "test-fixtures/lambdatest.zip"
-  runtime_version      = "syn-1.0"
+  runtime_version      = "syn-nodejs-puppeteer-3.2"
 
   schedule {
     expression = "rate(0 minute)"
@@ -1035,7 +1035,7 @@ resource "aws_synthetics_canary" "test" {
   execution_role_arn   = aws_iam_role.test.arn
   handler              = "exports.handler"
   zip_file             = "test-fixtures/lambdatest.zip"
-  runtime_version      = "syn-1.0"
+  runtime_version      = "syn-nodejs-puppeteer-3.2"
 
   schedule {
     expression = "rate(0 minute)"
@@ -1059,7 +1059,7 @@ resource "aws_synthetics_canary" "test" {
   execution_role_arn   = aws_iam_role.test.arn
   handler              = "exports.handler"
   zip_file             = "test-fixtures/lambdatest.zip"
-  runtime_version      = "syn-1.0"
+  runtime_version      = "syn-nodejs-puppeteer-3.2"
 
   schedule {
     expression = "rate(0 minute)"
@@ -1080,7 +1080,7 @@ resource "aws_synthetics_canary" "test" {
   execution_role_arn   = aws_iam_role.test.arn
   handler              = "exports.handler"
   zip_file             = "test-fixtures/lambdatest.zip"
-  runtime_version      = "syn-1.0"
+  runtime_version      = "syn-nodejs-puppeteer-3.2"
 
   schedule {
     expression = "rate(0 minute)"

--- a/aws/resource_aws_synthetics_canary_test.go
+++ b/aws/resource_aws_synthetics_canary_test.go
@@ -16,6 +16,7 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
 	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
 	"github.com/terraform-providers/terraform-provider-aws/aws/internal/service/synthetics/finder"
+	"github.com/terraform-providers/terraform-provider-aws/aws/internal/tfresource"
 )
 
 func init() {
@@ -529,14 +530,17 @@ func testAccCheckAwsSyntheticsCanaryDestroy(s *terraform.State) error {
 			continue
 		}
 
-		name := rs.Primary.ID
-		_, err := finder.CanaryByName(conn, name)
+		_, err := finder.CanaryByName(conn, rs.Primary.ID)
+
+		if tfresource.NotFound(err) {
+			continue
+		}
+
 		if err != nil {
-			if isAWSErr(err, synthetics.ErrCodeResourceNotFoundException, "") {
-				return nil
-			}
 			return err
 		}
+
+		return fmt.Errorf("Synthetics Canary %s still exists", rs.Primary.ID)
 	}
 
 	return nil
@@ -546,22 +550,22 @@ func testAccCheckAwsSyntheticsCanaryExists(n string, canary *synthetics.Canary) 
 	return func(s *terraform.State) error {
 		rs, ok := s.RootModule().Resources[n]
 		if !ok {
-			return fmt.Errorf("synthetics Canary not found: %s", n)
+			return fmt.Errorf("Not found: %s", n)
 		}
 
 		if rs.Primary.ID == "" {
-			return fmt.Errorf("synthetics Canary name not set")
+			return fmt.Errorf("No Synthetics Canary ID is set")
 		}
 
-		name := rs.Primary.ID
 		conn := testAccProvider.Meta().(*AWSClient).syntheticsconn
 
-		out, err := finder.CanaryByName(conn, name)
+		output, err := finder.CanaryByName(conn, rs.Primary.ID)
+
 		if err != nil {
-			return fmt.Errorf("syntherics Canary %s not found in AWS", name)
+			return err
 		}
 
-		*canary = *out.Canary
+		*canary = *output
 
 		return nil
 	}

